docs/features/source-generators.md

```diff
@@ -0,0 +1,285 @@
+# Source Generators
+
+
+## Summary
+
+Source generators aim to enable _compile time metaprogramming_, that is, code that can be created
+at compile time and added to the compilation. Source generators will be able to read the contents of 
+the compilation before running, as well as access any _additional files_, enabling generators to
+introspect both user C# code and generator specific files.
+
+> **Note**: This proposal is separate from the [previous generator design](generators.md)
+
+### High Level Design Goals
+
+- Generators produce one or more strings that represent C# source code to be added to the compilation.
+- Explicitly _additional_ only. Generators can add new code to a compilation but may **not** modify existing user code.
+- May access _additional files_, that is, non-C# source texts.
+- Run _un-ordered_, each generator will see the same input compilation.
+- A user specifies the generators to run via list of assemblies, much like analyzers.
+
+## Implementation
+
+At the simplest level source generators are an implementation of `Microsoft.CodeAnalysis.ISourceGenerator`
+
+```csharp
+namespace Microsoft.CodeAnalysis
+{
+    public interface ISourceGenerator
+    {
+        void Execute(SourceGeneratorContext context);
```

Comment:
synchronous?  non-cancellable?  how does that design impact compilations?

Reply:
> non-cancellable?

`context.CancellationToken`

Reply:
Generators very frequently need to read from the disk, particularly if their input is a mix of the semantic model and external files. Allowing for async may help for performance ?

Reply:
@jeromelaban 

Generators that read from disk in the context of source generators are an anti-pattern that must be avoided. To properly fit into our build model all inputs to the compiler, and by consequence generators, must be visible to MSBuild. If they're not then changes to those files will not result in compilation being re-done but rather MSBuild considering the compilation up to date and hence skipping it entirely.

Source generators that want to read non-source files must do so through the `/additionalFiles` options. This is the same approach that we take for analyzers.

Reply:
@jaredpar Thanks! I do not understand the rationale behind the anti-pattern. Reading files from the disk and performing up-to-date checks are two separate tasks. Could you explain a bit more about it?

The additional files option is also very limiting in its current form. While it provides access to the files content, it removes access to metadata provided by msbuild (most notably `Link` or others like nuget source) which can be very useful in some situations. `AdditionalText` also a synchronous API, which may force generators to spawn tasks to read content when there are many files (think many hundreds) and do parallel work when needed.

Reading through additional files doc, it does have the ability to tie a specific set of input files to a specific generator, which will force to rerun all generators. For a first version, it may be good enough, or could be done through msbuild metadata.

One information that could prove very useful for additional files is the diff with a previous run with the same compilation instance. Generators could use it to partially generate source or content, or reuse previously generated content. Diffing of compilations may not be useful at all, as determining what's changed could be challenging.

Reply:
> One information that could prove very useful for additional files is the diff with a previous run with the same compilation instance. 

That would require state tracking, which could easily lead to broken behavior.  

> also a synchronous API, which may force generators to spawn tasks to read content when there are many files (think many hundreds) and do parallel work when needed.

This is precisely the sort of perf cliff that terrifies me.  VS performance has already nose-dived in recent releases for me, and we basically had to stop using analyzers over here because of the massive cost they added to a build. This just seems to be exacerbating the issue.

Reply:
> Thanks! I do not understand the rationale behind the anti-pattern. Reading files from the disk and performing up-to-date checks are two separate tasks. Could you explain a bit more about it?

If you need to read in contents of a file, then you need to specify that file as a dependency.  Otherwise you simply will not be run if that file does not change.  Adding that file as a dependency can be done with .AdditionalFiles.  This informs the entire system about this dependency so that it can properly be used when computing what actually needs to run/update during a build.

Reply:
Performance is definitely a concern, and having generators running at every keystroke is scenario that most likely won't scale. For instance, it is frequent that some complex generators can take 15+ seconds to run, though part of it is just creating the initial `Compilation` that roslyn already has.

Running generators at build time, then refresh intellisense afterwards is a scenario that can be reasoned with pretty easily (this is what Uno.SourceGeneration does at this point), once explained.

Reply:
> Reading files from the disk and performing up-to-date checks are two separate tasks. Could you explain a bit more about it?

The way MSBuild evaluates up to date checks is to essentially analyze the inputs to Targets, check to see if they're unchanged and if so not run the target. That means for the compiler to properly participate in up to date checks all of the files used by the compiler must be listed in the Targets. This applies to both the core compiler and any analyzer or source generator plugin. 

The `/additionalFiles` parameter is how we allow analyzers, and now source generators, participate into this process.

Further an at API level the compiler must be able to represent an immutable snapshot of a compilation. This is a core part of our model and that includes removing the state of the file system from the equation. Hence all file inputs to the compiler end up getting represented as `SourceText` instances including the input to `/additionalFiles`. 

If source generators, or analyzers, directly read from disk they're violating at least the immutable snapshot property of the model and / or the MSBuild targets model.